File: zkmax/src/org/zkoss/zkmax/zul/api/Tablelayout.java
/* Tablelayout.java

{{IS_NOTE
	Purpose:
		
	Description:
		
	History:
		Tue Oct 23 09:22:13     2008, Created by Flyworld
}}IS_NOTE

Copyright (C) 2008 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
	This program is distributed under GPL Version 3.0 in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.
}}IS_RIGHT
 */
package org.zkoss.zkmax.zul.api;

import org.zkoss.zk.ui.WrongValueException;

/**
 * Tablelayout lay outs a container as an HTML table whose columns can be
 * specified, and rowspan and colspan of its child can also be specified to
 * create complex layouts within the table.
 * 
 * <p>
 * Default {@link org.zkoss.zkmax.zul.Tablelayout#getZclass}: z-table-layout.
 * 
 * @author robbiecheng
 * @since 3.5.0
 */
public interface Tablelayout extends org.zkoss.zul.impl.api.XulElement {

	/**
	 * Sets the number of columns.
	 */
	public void setColumns(int columns) throws WrongValueException;

	/**
	 * Returns number of rows to span this header. Default: 1.
	 */
	public int getColumns();

}

File: zkex/src/archive/web/js/zkex/wgt/mold/tablelayout.js
/* tablelayout.js

	Purpose:
		
	Description:
		
	History:
		Thu May 19 14:17:24     2009, Created by kindalu

Copyright (C) 2009 Potix Corporation. All Rights Reserved.

This program is distributed under GPL Version 3.0 in the hope that
it will be useful, but WITHOUT ANY WARRANTY.
*/
function (out) {
	var zcls = this.getZclass();
	var curRow=0, 
		curCol=0,
		rowspan,
		colspan,
		totalCol=this.getColumns(),
		colSpanList={};
	
	out.push('<table', this.domAttrs_(),'>');
		for(var child = this.firstChild; child; child = child.nextSibling){
			if(curCol == 0 || (colSpanList[curRow] && colSpanList[curRow] == curCol))
				out.push("<tr>");
				
			child.redraw(out);
			
			rowspan = child.getRowspan();
			colspan = child.getColspan();
			
			if(rowspan > 1){
				for(var row = curRow+1; row < curRow+rowspan; row++){
					if(!colSpanList[row])
						colSpanList[row] = colspan;
					else
						colSpanList[row] += colspan;
				}
			}
			
			if(curCol + colspan >= totalCol){
				curRow++;
				if(colSpanList[curRow])
					curCol = colSpanList[curRow];
				else
					curCol = 0;
				
				out.push("</tr>");
			}else{
				curCol += colspan;
			}
		}
		
		if(curCol!=0 && !(colSpanList[curRow] && colSpanList[curRow] == curCol) && curCol<totalCol)
			out.push("</tr>");
	out.push('</table>');
}